src/bosh-google-cpi/action/attach_disk.go

```diff
@@ -38,26 +40,56 @@ func (ad AttachDisk) Run(vmCID VMCID, diskCID DiskCID) (interface{}, error) {
 		return nil, api.NewDiskNotFoundError(string(diskCID), false)
 	}
 
+	// If this disk isn't already attached to this VM at the IAAS level, do that now
+	if !diskAttachedToVM(vmCID, disk.Users) {
+		_, err := ad.gceAttach(vmCID, diskCID, disk.SelfLink)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// The disk may now be configured by the BOSH agent
+	dev, err := ad.vmService.DiskDetail(string(vmCID), disk.SelfLink)
+	if err != nil {
+		return nil, bosherr.WrapErrorf(err, "Retrieving disk detail before BOSH agent attach")
+	}
+	err = ad.agentAttach(vmCID, diskCID, dev)
+	return nil, err
+}
+
+func (ad AttachDisk) gceAttach(vmCID VMCID, diskCID DiskCID, diskSelfLink string) (instance.DiskAttachmentDetail, error) {
 	// Atach the Disk to the VM
-	deviceName, devicePath, err := ad.vmService.AttachDisk(string(vmCID), disk.SelfLink)
+	dev, err := ad.vmService.AttachDisk(string(vmCID), diskSelfLink)
 	if err != nil {
 		if _, ok := err.(api.CloudError); ok {
-			return nil, err
+			return dev, err
```

Comment:
It appears the call-site for gceAttach() throws away the return value in the error case, perhaps returning nil would still be appropriate here.

Reply:
Good catch!

Reply:
To return `nil`, they type will need to be a pointer. Worth it? Or empty struct (`DiskAttachmentDetail{}`) for clarity?

Reply:
Empty struct would be clearer IMHO.

Reply:
If you'd prefer to not change the function's signature I mean.

Reply:
Signature changed.